---
title: ""
resource_files:
- dashboard_data/clinical.rds
- dashboard_data/kegg.rds
- dashboard_data/LD_ERPRneg.rds
- dashboard_data/LD_overall.rds
- dashboard_data/modules.rds
- dashboard_data/neighbor_ERPRneg.rds
- dashboard_data/neighbor_overall.rds
- dashboard_data/neighbors.rds
- dashboard_data/nmol.rds
- dashboard_data/oncosig.rds
- dashboard_data/pathways.rds
- dashboard_data/varComp_standardized_July27.rds
- dashboard_data/tf_spca_standardized_July27.rds
- dashboard_data/mirna_spca_standardized_July27.rds
- dashboard_data/tf_spca_key_standardized_July27.rds
- dashboard_data/mirna_spca_key_standardized_July27.rds
- dashboard_data/corlist_July27.rds
- dashboard_data/cancerabbrev.rds
- dashboard_data/refGene_keep.txt
- edge-dashboard-functions.R
runtime: shiny
output:
  flexdashboard::flex_dashboard:
    orientation: rows
    social: menu
    theme: bootstrap
---

```{r global, include=FALSE}
library(flexdashboard)
library(rvest)
library(scales)
library(pheatmap)
library(RColorBrewer)
library(data.table)
library(ggplot2)
library(plotly)
library(DT)
library(tidyr)
library(dplyr)
library(viridis)
library(mixOmics)
library(ComplexHeatmap)
library(circlize)

for(a in c("pathways", "kegg", "neighbors", "modules", "oncosig",
           "LD_overall", "LD_ERPRneg", "neighbor_ERPRneg", "neighbor_overall",
           "clinical", "nmol")) {
  cat(a,"\n")
  tmp <- readRDS(paste0("dashboard_data/", a, ".rds"))
  assign(a, tmp)
}

rownames(nmol) <- c("rnaseq", "mirna", "cna", "methyl", "mut", "samples", "mirna_ga")
all_values <- readRDS("dashboard_data/varComp_standardized_July27.rds") ## Memory 53, time 660
#all_values <- fread("dashboard_data/varComp.csv") ## Memory 51, time 3620
#all_values <- read.table("dashboard_data/varComp.csv", nrows=501152, 
#                               stringsAsFactors=FALSE, ## Memory 99, time 3010 
#                               check.names=FALSE,
#                               colClasses=c(rep("character", 2), rep("numeric", 12)), sep=",") 
#all_values <- read.csv.raw("dashboard_data/varComp.csv") ## Memory 195, time 2460

TF_values <- readRDS("dashboard_data/tf_spca_standardized_July27.rds")
mir_values <- readRDS("dashboard_data/mirna_spca_standardized_July27.rds")
TF_key<- readRDS("dashboard_data/tf_spca_key_standardized_July27.rds")
mir_key <- readRDS("dashboard_data/mirna_spca_key_standardized_July27.rds")

cancerabbrev <- readRDS("dashboard_data/cancerabbrev.rds")
cor_list <- readRDS("dashboard_data/corlist_July27.rds")

cancer_choice <- sort(unique(all_values$cancer))
names(cancer_choice) <- cancerabbrev$V2[match(cancer_choice, cancerabbrev$V1)]
gene_names <- unique(all_values$gene)

refgene <- fread("dashboard_data/refGene_keep.txt")
all_values <- merge(all_values, refgene, all.X=TRUE, by="gene")

source("edge-dashboard-functions.R")
```

Welcome
===================================== 


<style type="text/css">
.chart-title {  /* chart_title  */
font-size: 25px;
font-family: Helvetica;
</style>


### Exploring drivers of gene expression in TCGA cancer genomes (EDGE in TCGA)
Andrea Rau and Paul L. Auer


<div style="width:450px; height=300px">
![](images/logos.png)
</div>
<br>

*EDGE in TCGA* is an application that facilitates an interactive exploration of the genetic and epigenetic drivers of gene expression in 17 different cancers included in The Cancer Genome Project (TCGA). By decomposing the variance of gene expression into parts attributable to methylation, copy number alterations, somatic mutations, miRNA abundance, transcription factor expression, and cis-genetic polymorphisms, we provide a powerful tool for exploring patterns of genome-wide regulation as well as shared and divergent regulatory mechanisms across a wide variety of cancers.

<br>

The results presented here are in whole or part based upon data generated by The Cancer Genome Atlas managed
by the NCI and NHGRI. Information about TCGA Research Network may be found at http://cancergenome.nih.gov.

<br>

**FAQ**

1. How do I use the app?

Tutorials, pipeline options, and detailed explanations will be added soon.

2. Which cancers are included?

There are a total of 17 cancers: Skin cutaneous melanoma (SKCM), Kidney renal clear cell carcinoma (KIRC), Thyroid carcinoma (THCA), Sarcoma (SARC), Pheochromocytoma and paraganglioma (PCPG), Pancreatic adenocarcinoma (PAAD), Esophageal carcinoma (ESCA), Kidney renal papillary cell carcinoma (KIRP), Lung    adenocarcinoma (LUAD), Brain lower grade glioma (LGG), Prostate adenocarcinoma (PRAD), Head and neck squamous cell carcinoma (HNSC), Liver hepatocellular carcinoma (LIHC), Bladder urothelial cancer (BLCA), Stomach adenocarcinoma (STAD), Breast invasive carcinoma (BRCA), Cervical squamous cell carcinoma and endocervical adenocarcinoma (CESC). 


3. Where can I find more information?

A preprint will be submitted to bioRxiv shortly. You can find additional information about TCGA at the Genomic Data Commons: https://gdc.cancer.gov.

4. Does EDGE in TCGA require login or registration?

No, login and/or registration are not required. EDGE in TCGA is open-access.

5. Who should I contact if I have questions about EDGE in TCGA?

Please contact [Dr. Andrea Rau](http://www.andrea-rau.com) (andrea.rau@inra.fr) or [Dr. Paul Livermore Auer](http://people.uwm.edu/pauer)  (pauer@uwm.edu) if you have questions or experience any technical difficulties with the app.




Genome-wide pan-cancer
=====================================
Inputs {.sidebar}
-----------------------------------------------------------------------

```{r}
radioButtons("pancancerviz", label = h4(""), 
             choices = c("Heatmap", "PCA plot"), 
             selected = "Heatmap")
```

<br>

```{r}
uiOutput("showheatmapoptions0")
uiOutput("showheatmapoptions")
uiOutput("showPCAoptions")
uiOutput("showheatmapoptions2")
```

<br>

```{r}
uiOutput("showheatmapoptions3")
uiOutput("showPCAoptions2")
uiOutput("showheatmapoptions4")
```

<br> 

```{r}
panvizchoice <- reactive(as.character(input$pancancerviz))
PCA_source <- reactive(as.character(input$PCA_source))
heatmap_setchoose2 <- reactive(as.character(input$heatmap_setchoose2))
heatmap_settype2 <- reactive(as.character(input$heatmap_settype2))
heatmap_molsource <- reactive(as.character(input$heatmap_molsource))
myownlist <- reactive(as.character(input$myownlist))


downloadHandler(filename = function() {
  ifelse(input$pancancerviz == "PCA plot", 
         paste0('genomewide-pancancer-PCA-', input$PCA_source, '.pdf', sep=''),
         paste0('genomewide-pancancer-heatmap-', 
                ifelse(input$heatmap_settype2 == "Top 200", "Top200", 
                       ifelse(input$heatmap_settype2 == "Input my own list...", "custom", 
                       gsub(".", "_", input$heatmap_setchoose2, fixed=TRUE))), 
                '.pdf', sep=''))
}, content = function(file) {
  if(input$pancancerviz == "PCA plot") {
    pdf(file, width=10, height=7)
    tmp <- all_values[,c("gene", "cancer", PCA_source()), with=FALSE]
    tmp2 <- spread(tmp, key="cancer", value=PCA_source()) %>% dplyr::select(-gene)
    tmp2 <- tmp2[which(rowSums(is.na(tmp2)) == 0),]
    pca_analysis <- pca(t(tmp2))
    plotIndiv(pca_analysis, title="", cex=5)
    dev.off()
  }
  if(input$pancancerviz == "Heatmap") {
    pdf(file, width=8, height=7)
   
  if(heatmap_settype2() != "Top 200" & heatmap_settype2() != "Input my own list...") {
      if(heatmap_settype2() == "Kegg pathways") ind <- unlist(kegg[[heatmap_setchoose2()]])
      if(heatmap_settype2() == "Cancer gene neighborhoods") ind <- unlist(neighbors[[heatmap_setchoose2()]])
      if(heatmap_settype2() == "Cancer modules") ind <- unlist(modules[[heatmap_setchoose2()]])
      if(heatmap_settype2() == "Oncogenic signatures") ind <- unlist(oncosig[[heatmap_setchoose2()]])
      top200 <- FALSE
  }
  if(heatmap_settype2() == "Top 200") {
    ind <- 1
    top200 <- TRUE
  }
  if(heatmap_settype2() == "Input my own list...") {
    ind <- NULL
    if(length(myownlist())) {
        ind <- strsplit(myownlist(), split=",", fixed=TRUE) %>% unlist()
    }
    top200 <- FALSE
  }
  set.seed(12345)
  h <- my_complexheatmap(all_values=all_values, ind=ind, 
                    molsource_choice = heatmap_molsource(),
                    top200=top200, rn=ifelse(heatmap_settype2() == "Input my own list...",
                                               TRUE, FALSE))
  print(h)
    
    dev.off()
  }},  contentType="application/pdf", outputArgs=list(label="Download plot"))
```


```{r}

output$showheatmapoptions0 <- renderUI({
  if(panvizchoice() == "Heatmap") {
    checkboxGroupInput("heatmap_molsource", label=h4("Source of molecular variation"),
                       choices = c("mut", "cna", "methyl", "mirna", "TF", "genetic"),
                       selected = c("mut", "cna", "methyl", "mirna", "TF", "genetic"))
  }
})
output$showheatmapoptions <- renderUI({
  if(panvizchoice() == "Heatmap") {
    selectInput("heatmap_settype2", label = h4("Gene set type"),
                choices = c("Top 200", names(pathways), "Input my own list..."), selected = "Top 200")
  }
})
output$showheatmapoptions2 <- renderUI({
  if(panvizchoice() == "Heatmap" & heatmap_settype2() != "Top 200" & heatmap_settype2() != "Input my own list...") {
    mytest <- as.character(input$heatmap_settype2)
    selectInput('heatmap_setchoose2', h4('Gene set'), pathways[[mytest]], 
                selected = pathways[[mytest]][1])
  }
})
output$showPCAoptions <- renderUI({
  if(panvizchoice() == "PCA plot") {
    selectInput("PCA_source", label = h6("Select a source of molecular variation to visualize:"),
            choices = c("mut", "cna", "methyl", "mirna", "TF", "genetic"),
            selected = "TF", multiple=FALSE)
    }
})
output$showPCAoptions2 <- renderUI({
  if(panvizchoice() == "PCA plot") {
    tags$div(class="header", checked=NA,
             tags$h6("This plot represents the first two principal components for the genomewide variance components from a user-selected source of molecular variation for each of the 17 cancers included in this study. This allows a visualization of how cancers cluster with one another according to global patterns of how gene expression variation is decomposed into different molecular components."))
    }
})

output$showheatmapoptions3 <- renderUI({
  if(panvizchoice() == "Heatmap" & heatmap_settype2() != "Top 200" & heatmap_settype2() != "Input my own list...") {
      website <- read_html(paste0("http://software.broadinstitute.org/gsea/msigdb/cards/",
                              heatmap_setchoose2()))
      temp <- html_nodes(website, ".lists4 td , th") %>% html_text()
      m <- matrix(c("Source", paste0("http://software.broadinstitute.org/gsea/msigdb/cards/",
                                 heatmap_setchoose2()), temp[1:26]), ncol=2, byrow=TRUE)
      colnames(m) <- c("Field", "Description")
      htmlOutput("heatmap_geneset_description")
 #     tags$div(class="header", checked=NA,
#               tags$h5("See 'Single cancer results' tab for more details."))
    }
})

output$showheatmapoptions4 <- renderUI({
  if(panvizchoice() == "Heatmap" & heatmap_settype2() == "Input my own list...") {
    textInput("myownlist", label=h6("Enter list of gene symbols separated by commas (e.g. PTPN14,A1BG,MYL2,GATA3)"),
             value="") 
  }
})

output$heatmap_geneset_description <- renderText({
  website <- read_html(paste0("http://software.broadinstitute.org/gsea/msigdb/cards/",
                              heatmap_setchoose2()))
  temp <- html_nodes(website, ".lists4 td , th") %>% html_text()
  m <- matrix(c("Source", paste0("http://software.broadinstitute.org/gsea/msigdb/cards/",
                                 heatmap_setchoose2()), temp[1:26]), ncol=2, byrow=TRUE)
  str1 <- "<i>Brief gene set description:</i>"
  str2 <- "<i>See \'Single cancer results\' tab for more details.</i>"
  str1b <-  m[4,2]
  HTML(paste(str1, str1b, str2, sep='<br/><br/>'))
})


```


Column 
-----------------------------------------------------------------------
```{r}
uiOutput("heatmap_vs_PCA1")
uiOutput("heatmap_vs_PCA2")

ind_reactive <- reactive({
  tmp <- TRUE
  if(heatmap_settype2() != "Top 200" & heatmap_settype2() != "Input my own list...") {
    if(heatmap_settype2() == "Kegg pathways") ind <- unlist(kegg[[heatmap_setchoose2()]])
    if(heatmap_settype2() == "Cancer gene neighborhoods") 
      ind <- unlist(neighbors[[heatmap_setchoose2()]])
    if(heatmap_settype2() == "Cancer modules") ind <- unlist(modules[[heatmap_setchoose2()]])
    if(heatmap_settype2() == "Oncogenic signatures") 
      ind <- unlist(oncosig[[heatmap_setchoose2()]])
    if(sum(ind %in% gene_names)==0) tmp <- FALSE
  }
  if(heatmap_settype2() == "Input my own list...") {
    ind <- NULL
    if(length(myownlist())) {
      ind <- strsplit(myownlist(), split=",", fixed=TRUE) %>% unlist()
    }
    if(sum(ind %in% gene_names) == 0) tmp <- FALSE
  }
  return(tmp)
})

output$heatmap_vs_PCA1 <- renderUI({
  if(panvizchoice() == "Heatmap") {
    if(ind_reactive() == TRUE) {
       plotOutput("panheatmapglobal", width=800, height=700)
    } else {
      h3(textOutput("pantext2"))
    }
  }
})
output$pantext2 <- renderText({ 
  "Please select a gene set to the left, or enter a comma-delimited list of valid gene symbols."
})

output$heatmap_vs_PCA2 <- renderUI({
  if(panvizchoice() == "PCA plot") {
    plotOutput("panPCA", width=1000, height=700)
  }
})

output$panheatmapglobal <- renderPlot({
  
      if(heatmap_settype2() != "Top 200" & heatmap_settype2() != "Input my own list...") {
      if(heatmap_settype2() == "Kegg pathways") ind <- unlist(kegg[[heatmap_setchoose2()]])
      if(heatmap_settype2() == "Cancer gene neighborhoods") ind <- unlist(neighbors[[heatmap_setchoose2()]])
      if(heatmap_settype2() == "Cancer modules") ind <- unlist(modules[[heatmap_setchoose2()]])
      if(heatmap_settype2() == "Oncogenic signatures") ind <- unlist(oncosig[[heatmap_setchoose2()]])
      top200 <- FALSE
  }
  if(heatmap_settype2() == "Top 200") {
    ind <- 1
    top200 <- TRUE
  }
  if(heatmap_settype2() == "Input my own list...") {
    ind <- NULL
    if(length(myownlist())) {
        ind <- strsplit(myownlist(), split=",", fixed=TRUE) %>% unlist()
    }
    top200 <- FALSE
  }
  set.seed(12345)
  if(length(heatmap_molsource()) & length(ind)) {
      my_complexheatmap(all_values=all_values, ind=ind, 
                    molsource_choice = heatmap_molsource(),
                    top200=top200, rn=ifelse(heatmap_settype2() == "Input my own list...",
                                               TRUE, FALSE))
  }
})

output$panPCA <- renderPlot({
  tmp <- all_values[,c("gene", "cancer", PCA_source()), with=FALSE]
  tmp2 <- spread(tmp, key="cancer", value=PCA_source()) %>% dplyr::select(-gene)
  tmp2 <- tmp2[which(rowSums(is.na(tmp2)) == 0),]
  pca_analysis <- pca(t(tmp2))
  pp <- plotIndiv(pca_analysis, title="", cex=5)$graph
  pp + theme(plot.margin=margin(t=0.5,r=0.5,l=0.5,b=1.5, unit="in"))
})

```



Gene-wise pan-cancer
===================================== 
Inputs {.sidebar}
-----------------------------------------------------------------------
```{r}
textInput("gene33", label = h3("Gene symbol"), 
          value = "PTPN14")
#checkboxInput("suggestions", label = h6("Show auto-complete suggestions?"), 
#              value = FALSE, width = NULL)
```
e.g., PTPN14, A1BG, MYL2, GATA3

```{r}
#actionButton("go", "Make pan-cancer plots!")
```
<br>
```{r}
selectInput("cancersub", label = h4("Cancer subset"),
            choices = as.character(cancer_choice),
            selected = cancer_choice, multiple=TRUE)

cancersubset <- reactive({as.character(input$cancersub)})

downloadHandler(filename = function() {paste('genewise-pancancer-', gene(), '.pdf', sep='')},
                content = function(file) {
                  pdf(file, width=10, height=6, onefile=TRUE)
                    p <- my_dotplot(all_values, gene_choice=gene()) 
                    print(p)
                    my_pheatmap(gene_selection = gene(), cancer_selection = cancersubset(), 
                                all_values=all_values, cancerabbrev=cancerabbrev, main = gene()) 
                     dev.off()
                }, 
                contentType="application/pdf", outputArgs=list(label="Download plots"))
```





Column {.tabset}
-----------------------------------------------------------------------
### Variance components heatmap {data-padding=25}
```{r}
gene <- reactive(as.character(input$gene33))

#suggestions <- reactive(as.integer(input$suggestions))
# # https://github.com/VizWizard/BoxPlotR.shiny/blob/master/server.R
# downloadHandler(filename = function() {paste('pancancer-', Sys.Date(), '.pdf', sep='')},
#                 content = function(file) {
#                   pdf(file, width=10, height=6, onefile=FALSE)
#                   my_pheatmap(gene_selection = gene(), cancer_selection = cancersubset(), 
#                               all_values=all_values, cancerabbrev=cancerabbrev, 
#                               main=paste0("Gene: ", gene()))
#                   dev.off()
#                   }, 
#                 contentType="application/pdf")
# renderPlot({
#   my_pheatmap(gene_selection = gene(), cancer_selection = cancersubset(), 
#               all_values=all_values, cancerabbrev=cancerabbrev) 
# }, res=105)


uiOutput("panheatmap")

output$panheatmap <- renderUI({
  # if(suggestions() == TRUE) {
  #   textOutput("sugg")
  # } else {
  if(!gene() %in% unique(all_values$gene)) {
    h3(textOutput("pantext"))
  } else {
    plotOutput("panheatmapplot")
  }
  # }
})

output$pantext <- renderText({ 
  "Please enter a valid gene symbol."
})
# output$sugg <- renderText({ 
#   if(nchar(gene()) < 2) print("Please keep typing...") else {
#       paste0(all_values$gene[which(startsWith(unique(all_values$gene), gene())==TRUE)],
#              collapse="; ")
#   }
# })
output$panheatmapplot <- renderPlot({
  my_pheatmap(gene_selection = gene(), cancer_selection = cancersubset(), 
              all_values=all_values, cancerabbrev=cancerabbrev, main = gene()) 
}, res=105)
```


### Variance components dotplot {data-padding=25}

```{r}
uiOutput("pandotplot")

output$pandotplot <- renderUI({
  if(!gene() %in% unique(all_values$gene)) {
    h3(textOutput("pantext"))
  } else {
    plotlyOutput("pandotplotly")
  }
})
output$pandotplotly <- renderPlotly({
  g <- my_dotplot(all_values, gene_choice=gene()) 
  g2 <- ggplotly(g + theme(legend.position="none"), tooltip=c("cancer", "value", "type"))
  # y <- list(title = "Weighted sPCA loading")
  g2 
})

```





TF/miRNA contribution
=====================================
Inputs {.sidebar}
-----------------------------------------------------------------------
```{r}
selectInput("cancerTF", label = h3("Cancer"),
            choices = as.vector(cancer_choice),
            selected = "BRCA")

textInput("geneTF", label = h3("Gene symbol"), 
          value = "PTPN14")
#checkboxInput("suggestions", label = h6("Show auto-complete suggestions?"), 
#              value = FALSE, width = NULL)
```
e.g., PTPN14, A1BG, MYL2, GATA3

```{r}
downloadHandler(filename = function() {paste('TF+miRNA-', geneTF(), '-', cancerTF(), '.pdf', sep='')},
                content = function(file) {
                  pdf(file, width=10, height=6, onefile=TRUE)
                    p1 <- my_zoom(gene_selection=geneTF(), cancer_selection=cancerTF(), 
                       values=TF_values, type="TF", horiz=TRUE)
                    p2 <- my_zoom(gene_selection=geneTF(), cancer_selection=cancerTF(), 
                       values=mir_values, type="mirs", horiz=TRUE)
                    print(p1)
                    print(p2)
                  dev.off()
                }, 
                contentType="application/pdf", outputArgs=list(label="Download TF/miRNA plots"))
```
<br> 

```{r}
downloadHandler(filename = function() {paste('TF+miRNA-', geneTF(), '-', cancerTF(), '.csv', sep='')},
                content = function(file) {
                  ## Gather together TF results
                  sel <- TF_values[which(TF_values$gene == geneTF() & TF_values$cancer == cancerTF()),]
                  sel_list <- setNames(split(sel, seq(nrow(sel))), sel$cancer)
                  sel_list <- lapply(sel_list, function(x) {
                    tmp <- x[1,3]
                    colnames(x)[-c(1:3)] <- strsplit(TF_key[as.numeric(tmp),2], ";")$tfs
                    return(x)}
                  )
                  sel_list[is.na(sel_list)] <- 0
                  sel_list <- rbindlist(sel_list)
                  sel_list_tidy <- gather(sel_list, key=TF, value=loading, -cancer, -key, -gene) %>%
                    dplyr::select(-cancer, -gene, -key)
                  sel_list_tidy_TF <- sel_list_tidy
                  colnames(sel_list_tidy_TF) <- c("TF_or_miRNA", "weighted_loading")
                  ## Gather together mir results
                  sel <- mir_values[which(mir_values$gene == genemir() & mir_values$cancer == cancermir()),]
                  sel_list <- setNames(split(sel, seq(nrow(sel))), sel$cancer)
                  sel_list <- lapply(sel_list, function(x) {
                      tmp <- x[1,2]
                      colnames(x)[-c(1:3)] <- strsplit(mir_key[as.numeric(tmp),2], ";")$mirs
                    return(x)}
                  )
                  sel_list[is.na(sel_list)] <- 0
                  sel_list <- rbindlist(sel_list)
                  sel_list_tidy <- gather(sel_list, key=miRNA, value=loading, -cancer, -key, -gene) %>%
                    dplyr::select(-cancer, -gene, -key)
                  colnames(sel_list_tidy) <- c("TF_or_miRNA", "weighted_loading")
                  write.csv(x= rbind(sel_list_tidy_TF, sel_list_tidy), file=file, quote=FALSE, row.names=FALSE)
                }, 
                contentType="text/csv", outputArgs=list(label="Download TF/miRNA tables"))
```


Column {.tabset}
-----------------------------------------------------------------------
### TF overview {data-padding=20}
```{r}
cancerTF <- reactive(as.character(input$cancerTF))
geneTF <- reactive(as.character(input$geneTF))

uiOutput("panTF")

output$panTF <- renderUI({
  #if(suggestions() == TRUE) {
  #   textOutput("suggTF")
  # } else {
  if(!geneTF() %in% unique(all_values$gene)) {
    h3(textOutput("pantextTF"))
  } else {
    plotlyOutput("panTFplot")
  }
  #  }
})

output$pantextTF <- renderText({ 
  "Please enter a valid gene symbol to begin."
})
output$suggTF <- renderText({ 
  if(nchar(geneTF()) < 2) print("Please keep typing...") else {
    paste0(all_values$gene[which(startsWith(unique(all_values$gene), geneTF())==TRUE)],
           collapse="; ")
  }
})
output$panTFplot <- renderPlotly({
  withProgress(message = 'Making plot...', value = 0, {
    p <- my_zoom(gene_selection=geneTF(), cancer_selection=cancerTF(), 
                 values=TF_values, type="TF", horiz=TRUE)
    p2 <- ggplotly(p + ylab(" "), tooltip=c("TF", "loading"))
    p2 
  })
  
})
```

### TF browse {data-padding=20}
```{r}
renderDataTable({
  sel <- TF_values[which(TF_values$gene == geneTF() & TF_values$cancer == cancerTF()),]
  sel_list <- setNames(split(sel, seq(nrow(sel))), sel$cancer)
  type <- "TF"
  sel_list <- lapply(sel_list, function(x) {
    if(type=="TF") {
      tmp <- x[1,3]
      colnames(x)[-c(1:3)] <- strsplit(TF_key[as.numeric(tmp),2], ";")$tfs
    }
    if(type=="mirs") {
      tmp <- x[1,2]
      colnames(x)[-c(1:3)] <- strsplit(mir_key[as.numeric(tmp),2], ";")$mirs
    }
    return(x)}
  )
  sel_list[is.na(sel_list)] <- 0
  sel_list <- rbindlist(sel_list)
  sel_list_tidy <- gather(sel_list, key=TF, value=loading, -cancer, -key, -gene) %>%
    dplyr::select(-cancer, -gene, -key)
  DT::datatable(sel_list_tidy, rownames=FALSE,
                filter=list(position="top", clear=TRUE, plain=FALSE),
                caption=paste0("Weighted sPCA loadings"))

})
```


### miRNA overview {data-padding=20}

```{r}
cancermir <- reactive(as.character(input$cancerTF))
genemir <- reactive(as.character(input$geneTF))

uiOutput("panmir")

output$panmir <- renderUI({
  # if(suggestions() == TRUE) {
  #   textOutput("suggmir")
  # } else {
  if(!genemir() %in% unique(all_values$gene)) {
    h3(textOutput("pantextmir"))
  } else {
    plotlyOutput("panmirplot")
  }
  # }
})

output$pantextmir <- renderText({ 
  "Please enter a valid gene symbol to begin."
})

output$suggmir <- renderText({ 
  if(nchar(genemir()) < 2) print("Please keep typing...") else {
    paste0(all_values$gene[which(startsWith(unique(all_values$gene), genemir())==TRUE)],
           collapse="; ")
  }
})

output$panmirplot <- renderPlotly({
  
  withProgress(message = 'Making plot...', value = 0, {
    
    p <- my_zoom(gene_selection=genemir(), cancer_selection=cancermir(), 
                 values=mir_values, type="mirs", horiz=TRUE)
    p2 <- ggplotly(p + ylab(" "), tooltip=c("miRNA", "loading"))
    p2
  })
  
})

```

### miRNA browse {data-padding=20}
```{r}
renderDataTable({
  sel <- mir_values[which(mir_values$gene == genemir() & mir_values$cancer == cancermir()),]
  sel_list <- setNames(split(sel, seq(nrow(sel))), sel$cancer)
  type <- "mirs"
  sel_list <- lapply(sel_list, function(x) {
    if(type=="TF") {
      tmp <- x[1,3]
      colnames(x)[-c(1:3)] <- strsplit(TF_key[as.numeric(tmp),2], ";")$tfs
    }
    if(type=="mirs") {
      tmp <- x[1,2]
      colnames(x)[-c(1:3)] <- strsplit(mir_key[as.numeric(tmp),2], ";")$mirs
    }
    return(x)}
  )
  sel_list[is.na(sel_list)] <- 0
  sel_list <- rbindlist(sel_list)
  sel_list_tidy <- gather(sel_list, key=miRNA, value=loading, -cancer, -key, -gene) %>%
    dplyr::select(-cancer, -gene, -key)
  DT::datatable(sel_list_tidy, rownames=FALSE,
                filter=list(position="top", clear=TRUE, plain=FALSE),
                caption=paste0("Weighted sPCA loadings"))

})
```

Single cancer results
=====================================
Inputs {.sidebar}
-----------------------------------------------------------------------

```{r}
selectInput("cancersingle", label = h3("Cancer"),
            choices = as.vector(cancer_choice),
            selected = "BRCA")

radioButtons("singlecancerchoice", label = h4(""), 
             choices = c("Browse results", "Gene set analysis"), 
             selected = "Browse results")
```

<br>

```{r}
uiOutput("showgenesetoptions")
uiOutput("setchoose2")
uiOutput("setchoose2_plot")
```


```{r}
singlecancerchoice <- reactive(as.character(input$singlecancerchoice))
setchoose2 <- reactive(as.character(input$setchoose2))
settype2 <- reactive(as.character(input$settype2))
cancer <- reactive(as.character(input$cancersingle))

output$showgenesetoptions <- renderUI({
  if(singlecancerchoice() == "Gene set analysis") {
    selectInput("settype2", label = h4("Gene set type"),
                choices = names(pathways), selected = "Kegg pathways")
  }
})
output$setchoose2 <- renderUI({
  if(singlecancerchoice() == "Gene set analysis") {
    mytest <- as.character(input$settype2)
    selectInput('setchoose2', h4('Gene set'), pathways[[mytest]])
  }
})
output$setchoose2_plot <- renderUI({})
```
<br>


```{r}
downloadHandler(filename = function() {
  ifelse(singlecancerchoice() == "Browse results", paste('varianceComponents-fullResults-', cancer(), '.csv', sep=''),
         paste('geneset-heatmap-', cancer(), '-', gsub(".", "_", input$setchoose2, fixed=TRUE), '.pdf', sep=''))
  },
                content = function(file) {
                  if(singlecancerchoice() == "Browse results") {
                    sel <- all_values[, c("cancer", "gene", "chr", "strand", "tss", "mut", "cna", "mirna", "TF", 
                                      "methyl", "genetic", "residual")]
                    sel <- sel[which(sel$cancer == cancer()),]
                    sel <- sel[,-1]
                    write.csv(x=sel, file=file, quote=FALSE, row.names=FALSE)
                  }
                  if(singlecancerchoice() != "Browse results") {
                    if(settype2() == "Kegg pathways") ind <- unlist(kegg[[setchoose2()]])
                    if(settype2() == "Cancer gene neighborhoods") ind <- unlist(neighbors[[setchoose2()]])
                    if(settype2() == "Cancer modules") ind <- unlist(modules[[setchoose2()]])
                    if(settype2() == "Oncogenic signatures") ind <- unlist(oncosig[[setchoose2()]])
                    pdf(file, width=10, height=6, onefile=FALSE)
                    my_pheatmap(gene_selection = ind, cancer_selection = cancer(),
                      all_values=all_values, cancerabbrev=cancerabbrev)
                    dev.off()
                  }
                })
```

<br>
<br>

```{r}
renderTable({
  # tmp <- "BRCA"
  tmp <- cancer()
  if(length(grep("_", tmp)) > 0) tmp <- unlist(strsplit(tmp, split="_", fixed=TRUE))[1]
  website <- paste0("http://www.cbioportal.org/study?id=", tolower(tmp), "_tcga_pub#summary")
  website2 <- paste0("http://firebrowse.org/?cohort=", tolower(tmp))
  
  
  Links <- c(paste0("<a href='",  website, "' target='_blank'>cBioPortal</a>"),
             paste0("<a href='",  website2, "' target='_blank'>Broad Firebrowse</a>"))
  data.frame(Links)
  #  m <- matrix(website)
  #  colnames(m) <- c("cBioPortal link")
  #  m
}, sanitize.text.function = function(x) x)
```


Column 
-----------------------------------------------------------------------
```{r}
uiOutput("browse_vs_genesets")

output$browse_vs_genesets <- renderUI({
  
  if(singlecancerchoice() == "Browse results") {
    fluidPage(
      tabsetPanel(
        tabPanel("Results", value = "sub3", dataTableOutput("browsetable")),
        tabPanel("Sample size", value = "sub4",  dataTableOutput("summarytable")),
        tabPanel("Clinical information", value = "sub5", verbatimTextOutput("clinicaltable"))
      )) 
  } else {
    fluidPage(
      tabsetPanel(
        tabPanel("Heatmap", value = "sub1", plotOutput("genesetheatmap", width=1000, height=700)),
        tabPanel("Description", value = "sub2", dataTableOutput("genesetdescription"))
      ))
  }
})

output$browsetable <- DT::renderDataTable({
  # sel <- all_values[cancer == "BRCA", list(gene, chr, tss, mut, cna, mirna, TF, methyl, genetic, residual)]
 #  sel <- all_values[cancer == cancer(), .(gene, chr, tss, mut, cna, mirna, TF, methyl, genetic, residual)]
  sel <- all_values[, c("cancer", "gene", "chr", "tss", "mut", "cna", "mirna", "TF", "methyl", 
                        "genetic", "residual")]
  sel <- sel[which(sel$cancer == cancer()),]

  ## Round to 3 decimals
  cols <- names(sel)[5:11]
  sel <- sel %>% mutate_at(cols, funs(round(.,3)))
  DT::datatable(sel[,-1], options=list(pageLength = 25), rownames=FALSE,
                filter=list(position="top", clear=TRUE, plain=FALSE),
                caption="Note: TSS locations shown in hg19 coordinates.")
})

output$summarytable <- DT::renderDataTable({
  tmp <- data.frame(Item = c("Cancer",
                             "Sample size", "RNA-seq", "Somatic non-silent mutations",
                             "Somatic copy number alterations", "Methylation",
                             "miRNA-seq"),
                    Description = c(cancer(),
                                    paste(nmol[which(rownames(nmol) == "samples"),
                                               which(colnames(nmol) == cancer())],
                                          "\n(complete Caucasian tumoral samples)"),
                                    paste(nmol[which(rownames(nmol) == "rnaseq"),
                                               which(colnames(nmol) == cancer())], "genes\n(nonzero values)"),
                                    paste(nmol[which(rownames(nmol) == "mut"),
                                               which(colnames(nmol) == cancer())],
                                          "genes\n(mutation observed in at least one individual)"),
                                    paste(nmol[which(rownames(nmol) == "cna"),
                                               which(colnames(nmol) == cancer())], 
                                          "genes\n(nonzero values)"),
                                    paste(nmol[which(rownames(nmol) == "methyl"),
                                               which(colnames(nmol) == cancer())],
                                          "representative probes in gene promoter regions"),
                                    paste(nmol[which(rownames(nmol) == "mirna"),
                                               which(colnames(nmol) == cancer())],
                                          "microRNAs\n(nonzero values)")),
                    Detail =  c(cancerabbrev[which(cancerabbrev$V1 == cancer()),2],
                                "", "Illumina HiSeq 2000 (RSEM values)",
                                "",
                                "Affymetrix Genome-Wide Human SNP Array 6.0 (aggregated at gene-level using CNTools)",
                                "Illumina Infinium HumanMethylation450 BeadChip",
                                "Illumina HiSeq 2000 / Illumina Genome Analyzer (RPMMM values)")
  )
  DT::datatable(tmp, options=list(pageLength = 25))
})

output$clinicaltable <- renderPrint({
  lapply(as.data.frame(clinical[[cancer()]]), summary)[-1]
})


output$genesetheatmap <- renderPlot({
  if(settype2() == "Kegg pathways") ind <- unlist(kegg[[setchoose2()]])
  if(settype2() == "Cancer gene neighborhoods") ind <- unlist(neighbors[[setchoose2()]])
  if(settype2() == "Cancer modules") ind <- unlist(modules[[setchoose2()]])
  if(settype2() == "Oncogenic signatures") ind <- unlist(oncosig[[setchoose2()]])
  
  my_pheatmap(gene_selection = ind, cancer_selection = cancer(),
              all_values=all_values, cancerabbrev=cancerabbrev)
})

output$genesetdescription <- DT::renderDataTable({
  website <- read_html(paste0("http://software.broadinstitute.org/gsea/msigdb/cards/",
                              setchoose2()))
  temp <- html_nodes(website, ".lists4 td , th") %>% html_text()
  m <- matrix(c("Source", paste0("http://software.broadinstitute.org/gsea/msigdb/cards/",
                                 setchoose2()), temp[1:26]), ncol=2, byrow=TRUE)
  colnames(m) <- c("Field", "Description")
  DT::datatable(m, options=list(pageLength = 25))
  
})

```





Acknowledgements
=====================================

This work was made possible by the availability of large-scale open-access data from The Cancer Genome Atlas, as well as the open-source programming language R. In particular, we have made use of several R packages to perform data analysis and construct this interactive dashboard, including the following:

- [mixOmics](http://mixomics.org)
- [shiny](https://shiny.rstudio.com)
- [flexdashboard](http://rmarkdown.rstudio.com/flexdashboard/index.html)
- [tidyverse](https://www.tidyverse.org)
- [plotly](https://plot.ly/d3-js-for-r-and-shiny-charts)
- [rvest](https://github.com/hadley/rvest)
- [pheatmap](https://cran.r-project.org/web/packages/pheatmap/index.html)
- [ComplexHeatmap](https://bioconductor.org/packages/release/bioc/html/ComplexHeatmap.html)
- [RColorBrewer](https://cran.r-project.org/web/packages/RColorBrewer/index.html)
- [viridis](https://cran.r-project.org/web/packages/viridis/vignettes/intro-to-viridis.html)
- [data.table](https://cran.r-project.org/web/packages/data.table/index.html)
- [DT](https://rstudio.github.io/DT)

<br>

Dr. Andrea Rau was supported by the [AgreenSkills+](https://www.agreenskills.eu) fellowship program, which received fudning from the EU's Seventh Framework Program under grant agreement FP7-60939 (AgreenSkills+ contract). We thank the [UWM High Performance Computing (HPC) Service](http://uwm.edu/hpc) for computing resources used in this work.

<br>

The EDGE in TCGA resource is available under the GNU public license (GPL-3). If you use any of the results or figures generated by this app in a publication or presentation, please cite our work: Rau et al. (2017) Exploring drivers of gene expression in The Cancer Genome Atlas.

##############################
## Misc old scripts

<!-- Browse -->
<!-- ===================================== -->
<!-- Inputs {.sidebar} -->
<!-- ----------------------------------------------------------------------- -->


<!-- ```{r} -->
<!-- selectInput("cancer", label = h3("Cancer"), -->
<!--     choices = as.vector(cancer_choice), -->
<!--     selected = "BRCA") -->

<!-- renderTable({ -->
<!--   # tmp <- "BRCA" -->
<!--   tmp <- cancer() -->
<!--   if(length(grep("_", tmp)) > 0) tmp <- unlist(strsplit(tmp, split="_", fixed=TRUE))[1] -->
<!--   website <- paste0("http://www.cbioportal.org/study?id=", tolower(tmp), "_tcga_pub#summary") -->
<!--   website2 <- paste0("http://firebrowse.org/?cohort=", tolower(tmp)) -->


<!--   References <- c(paste0("<a href='",  website, "' target='_blank'>cBioPortal</a>"), -->
<!--                   paste0("<a href='",  website2, "' target='_blank'>Broad Firebrowse</a>")) -->
<!--   data.frame(References) -->
<!-- #  m <- matrix(website) -->
<!-- #  colnames(m) <- c("cBioPortal link") -->
<!-- #  m -->
<!-- }, sanitize.text.function = function(x) x) -->
<!-- ``` -->



<!-- Column -->
<!-- ----------------------------------------------------------------------- -->
<!-- ### {data-padding=20} -->
<!-- ```{r} -->
<!-- cancer <- reactive(as.character(input$cancer)) -->

<!-- downloadHandler(filename = function() {paste('datadownload-', Sys.Date(), '.csv', sep='')}, -->
<!--                 content = function(file) { -->
<!--   sel <- all_values[cancer == cancer(), .(gene, chr, strand, tss, mut, cna, mirna, TF, methyl, genetic, residual)] -->
<!--   write.csv(x=sel, file=file, quote=FALSE, row.names=FALSE) -->
<!--                   }, contentType="text/csv") -->

<!-- DT::renderDataTable({ -->
<!--   sel <- all_values[cancer == cancer(), .(gene, chr, tss, mut, cna, mirna, TF, methyl, genetic, residual)] -->
<!--   ## Round to 3 decimals -->
<!--   cols <- names(sel)[4:10] -->
<!--   sel <- sel %>% mutate_each_(funs(round(.,3)), cols) -->
<!--   DT::datatable(sel, options=list(pageLength = 10), rownames=FALSE,  -->
<!--                 filter=list(position="top", clear=TRUE, plain=FALSE), -->
<!--                             caption="Note: TSS locations shown in hg19 coordinates.") -->
<!-- }, res=105, width=1200, height=650) -->

<!-- ``` -->




<!-- Gene set analysis -->
<!-- ===================================== -->
<!-- Inputs {.sidebar} -->
<!-- ----------------------------------------------------------------------- -->


<!-- ```{r} -->
<!-- selectInput("cancer", label = h3("Cancer"), -->
<!--     choices = as.vector(cancer_choice), -->
<!--     selected = "BRCA") -->
<!-- ``` -->



<!-- ```{r} -->
<!--   selectInput("settype", label = h4("Gene set type"), -->
<!--             choices = names(pathways)) -->
<!--   uiOutput("setchoose") -->
<!-- ``` -->


<!-- Column {.tabset .tabset-fade} -->
<!-- ----------------------------------------------------------------------- -->

<!-- ### Variance components {data-padding=20} -->

<!-- ```{r} -->
<!-- settype <- reactive(as.character(input$settype)) -->
<!-- output$setchoose <- renderUI({ -->
<!--   mytest <- as.character(input$settype) -->
<!--   selectInput('setchoose', h4('Gene set'), pathways[[mytest]]) -->
<!-- }) -->
<!-- setchoose <- reactive(as.character(input$setchoose)) -->

<!-- downloadHandler(filename = function() {paste('pathways-', Sys.Date(), '.pdf', sep='')}, -->
<!--                 content = function(file) { -->
<!--                   if(settype() == "Kegg pathways") ind <- unlist(kegg[[setchoose()]]) -->
<!--                   if(settype() == "Cancer gene neighborhoods") ind <- unlist(neighbors[[setchoose()]]) -->
<!--                   if(settype() == "Cancer modules") ind <- unlist(modules[[setchoose()]]) -->
<!--                   if(settype() == "Oncogenic signatures") ind <- unlist(oncosig[[setchoose()]]) -->

<!--                   pdf(file, width=10, height=6, onefile=FALSE) -->
<!--                   my_pheatmap(gene_selection = ind, cancer_selection = cancer(),  -->
<!--                               all_values=all_values, cancerabbrev=cancerabbrev,  -->
<!--                               main=paste0(settype(), ": ", setchoose(), " (", cancer(), ")")) -->
<!--                   dev.off() -->
<!--                   },  -->
<!--                 contentType="application/pdf") -->

<!-- renderPlot({ -->
<!--   if(settype() == "Kegg pathways") ind <- unlist(kegg[[setchoose()]]) -->
<!--   if(settype() == "Cancer gene neighborhoods") ind <- unlist(neighbors[[setchoose()]]) -->
<!--   if(settype() == "Cancer modules") ind <- unlist(modules[[setchoose()]]) -->
<!--   if(settype() == "Oncogenic signatures") ind <- unlist(oncosig[[setchoose()]]) -->

<!--   my_pheatmap(gene_selection = ind, cancer_selection = cancer(),  -->
<!--               all_values=all_values, cancerabbrev=cancerabbrev)  -->

<!-- }, res=105, width=1200, height=650) -->

<!-- ``` -->

<!-- ### Pathway description -->

<!-- ```{r} -->
<!-- renderTable({ -->
<!--   website <- read_html(paste0("http://software.broadinstitute.org/gsea/msigdb/cards/", -->
<!--     setchoose())) -->
<!--   temp <- html_nodes(website, ".lists4 td , th") %>% html_text() -->
<!--   m <- matrix(c("Source", paste0("http://software.broadinstitute.org/gsea/msigdb/cards/", -->
<!--     setchoose()), temp[1:26]), ncol=2, byrow=TRUE) -->
<!--   colnames(m) <- c("Field", "Description") -->
<!--   m -->
<!-- }) -->
<!-- ``` -->



<!-- Details -->
<!-- ===================================== -->
<!-- Inputs {.sidebar} -->
<!-- ----------------------------------------------------------------------- -->


<!-- ```{r} -->
<!-- selectInput("cancer2", label = h3("Cancer"), -->
<!--     choices = as.vector(cancer_choice), -->
<!--     selected = "BRCA") -->

<!-- renderTable({ -->
<!--   # tmp <- "BRCA" -->
<!--   tmp <- cancer2() -->
<!--   if(length(grep("_", tmp)) > 0) tmp <- unlist(strsplit(tmp, split="_", fixed=TRUE))[1] -->
<!--   website <- paste0("http://www.cbioportal.org/study?id=", tolower(tmp), "_tcga_pub#summary") -->
<!--   website2 <- paste0("http://firebrowse.org/?cohort=", tolower(tmp)) -->


<!--   References <- c(paste0("<a href='",  website, "' target='_blank'>cBioPortal</a>"), -->
<!--                   paste0("<a href='",  website2, "' target='_blank'>Broad Firebrowse</a>")) -->
<!--   data.frame(References) -->
<!-- #  m <- matrix(website) -->
<!-- #  colnames(m) <- c("cBioPortal link") -->
<!-- #  m -->
<!-- }, sanitize.text.function = function(x) x) -->
<!-- ``` -->


<!-- Column {.tabset .tabset-fade} -->
<!-- ----------------------------------------------------------------------- -->

<!-- ### Summary {data-padding=20} -->

<!-- ```{r} -->
<!-- cancer2 <- reactive(as.character(input$cancer2)) -->

<!-- renderTable({ -->

<!--   tmp <- data.frame(Item = c("Cancer", -->
<!--                              "Sample size", "RNA-seq", "Somatic non-silent mutations", -->
<!--                              "Somatic copy number alterations", "Methylation", -->
<!--                              "miRNA-seq"), -->
<!--                     Description = c(cancer2(), -->
<!--                                     paste(nmol[which(rownames(nmol) == "samples"), -->
<!--                                          which(colnames(nmol) == cancer2())], -->
<!--                                          "\n(complete Caucasian tumoral samples)"), -->
<!--                                     paste(nmol[which(rownames(nmol) == "rnaseq"), -->
<!--                                          which(colnames(nmol) == cancer2())], "genes\n(nonzero values)"), -->
<!--                                     paste(nmol[which(rownames(nmol) == "mut"), -->
<!--                                          which(colnames(nmol) == cancer2())], -->
<!--                                          "genes\n(mutation observed in at least one individual)"), -->
<!--                                     paste(nmol[which(rownames(nmol) == "cna"), -->
<!--                                          which(colnames(nmol) == cancer2())],  -->
<!--                                          "genes\n(nonzero values)"), -->
<!--                                     paste(nmol[which(rownames(nmol) == "methyl"), -->
<!--                                          which(colnames(nmol) == cancer2())], -->
<!--                                          "representative probes in gene promoter regions"), -->
<!--                                     paste(nmol[which(rownames(nmol) == "mirna"), -->
<!--                                          which(colnames(nmol) == cancer2())], -->
<!--                                          "microRNAs\n(nonzero values)")), -->
<!--                     Detail =  c(cancerabbrev[which(cancerabbrev$V1 == cancer2()),2], -->
<!--                                 "", "Illumina HiSeq 2000 (RSEM values)", -->
<!--                                   "", -->
<!--           "Affymetrix Genome-Wide Human SNP Array 6.0 (aggregated at gene-level using CNTools)", -->
<!--           "Illumina Infinium HumanMethylation450 BeadChip", -->
<!--           "Illumina HiSeq 2000 / Illumina Genome Analyzer (RPMMM values)") -->
<!--   ) -->
<!--  tmp -->
<!-- }) -->

<!-- ``` -->

<!-- ### Clinical information {data-padding=20} -->

<!-- ```{r} -->
<!-- renderPrint({ -->
<!--    lapply(as.data.frame(clinical[[cancer2()]]), summary)[-1] -->
<!-- }, width=120) -->
<!-- ``` -->







<!-- BRCA -->
<!-- ===================================== -->

<!-- Inputs {.sidebar} -->
<!-- ----------------------------------------------------------------------- -->
<!-- ```{r} -->
<!--   selectInput("genelist", label = h3("Gene list"), -->
<!--             choices = c("Genes in LD (>0.6), BC overall", -->
<!--                         "Genes within 0.5 Mb, BC overall", -->
<!--                         "Genes in LD (>0.6), ER-PR neg", -->
<!--                         "Genes within 0.5 Mb, ER-PR neg")) -->
<!--   checkboxInput("dendrogram", label="Row dendrogram", TRUE) -->
<!-- ``` -->


<!-- Column {.tabset data-padding=20} -->
<!-- ------------------------------------- -->
<!-- ### Triple negative -->

<!-- ```{r} -->
<!-- genelist <- reactive({ -->
<!--   genelist_choice <- as.character(input$genelist) -->
<!--   if(genelist_choice == "Genes in LD (>0.6), BC overall") tmp <- LD_overall -->
<!--   if(genelist_choice == "Genes in LD (>0.6), ER-PR neg")  tmp <- LD_ERPRneg -->
<!--   if(genelist_choice == "Genes within 0.5 Mb, BC overall") tmp <- neighbor_overall -->
<!--   if(genelist_choice == "Genes within 0.5 Mb, ER-PR neg")  tmp <- neighbor_ERPRneg -->
<!--   tmp -->
<!-- }) -->
<!-- dendrogram <- reactive({ -->
<!--   ifelse(as.character(input$dendrogram) == "TRUE", TRUE, FALSE) -->
<!-- }) -->


<!-- renderD3heatmap({ -->
<!--  #  sel <- all_values[gene %in% LD_overall & cancer == "BRCA_tripleneg", .(gene, mut, cna, mirna, TF, methyl, genetic, residual)] -->
<!--   sel <- all_values[gene %in% genelist() & cancer == "BRCA_tripleneg", .(gene, mut, cna, mirna, TF, methyl, genetic, residual)] -->
<!--   sel <- as.data.frame(sel) -->
<!--   rownames(sel) <- sel$gene -->
<!--   pve_values <- sel[,-1] -->
<!--   d3heatmap(pve_values, colors="Reds", Colv=FALSE,  yaxis_font_size = "9pt", Rowv=dendrogram()) -->
<!-- }) -->
<!-- ``` -->

<!-- ### Luminal A -->
<!-- ```{r} -->
<!-- renderD3heatmap({ -->
<!--    #  sel <- all_values[gene %in% LD_overall & cancer == "BRCA_luminalA", .(gene, mut, cna, mirna, TF, methyl, genetic, residual)] -->
<!--   sel <- all_values[gene %in% genelist() & cancer == "BRCA_luminalA", .(gene, mut, cna, mirna, TF, methyl, genetic, residual)] -->
<!--   sel <- as.data.frame(sel) -->
<!--   rownames(sel) <- sel$gene -->
<!--   pve_values <- sel[,-1] -->
<!--   d3heatmap(pve_values, colors="Reds", Colv=FALSE,  yaxis_font_size = "9pt", Rowv=dendrogram()) -->
<!-- }) -->
<!-- ``` -->

<!-- ### HER2 negative -->
<!-- ```{r} -->
<!-- renderD3heatmap({ -->
<!--  #  sel <- all_values[gene %in% LD_overall & cancer == "BRCA_HER2neg", .(gene, mut, cna, mirna, TF, methyl, genetic, residual)] -->
<!--   sel <- all_values[gene %in% genelist() & cancer == "BRCA_HER2neg", .(gene, mut, cna, mirna, TF, methyl, genetic, residual)] -->
<!--   sel <- as.data.frame(sel) -->
<!--   rownames(sel) <- sel$gene -->
<!--   pve_values <- sel[,-1] -->
<!--   d3heatmap(pve_values, colors="Reds", Colv=FALSE,  yaxis_font_size = "9pt", Rowv=dendrogram()) -->
<!-- }) -->
<!-- ``` -->

<!-- ### HER2 positive -->
<!-- ```{r} -->
<!-- renderD3heatmap({ -->
<!--  #  sel <- all_values[gene %in% LD_overall & cancer == "BRCA_HER2pos", .(gene, mut, cna, mirna, TF, methyl, genetic, residual)] -->
<!--   sel <- all_values[gene %in% genelist() & cancer == "BRCA_HER2pos", .(gene, mut, cna, mirna, TF, methyl, genetic, residual)] -->
<!--   sel <- as.data.frame(sel) -->
<!--   rownames(sel) <- sel$gene -->
<!--   pve_values <- sel[,-1] -->
<!--   d3heatmap(pve_values, colors="Reds", Colv=FALSE,  yaxis_font_size = "9pt", Rowv=dendrogram()) -->
<!-- }) -->
<!-- ``` -->

<!-- ### BRCA overall -->

<!-- ```{r} -->
<!-- renderD3heatmap({ -->
<!--  #  sel <- all_values[gene %in% LD_overall & cancer == "BRCA", .(gene, mut, cna, mirna, TF, methyl, genetic, residual)] -->
<!--   sel <- all_values[gene %in% genelist() & cancer == "BRCA", .(gene, mut, cna, mirna, TF, methyl, genetic, residual)] -->
<!--   sel <- as.data.frame(sel) -->
<!--   rownames(sel) <- sel$gene -->
<!--   pve_values <- sel[,-1] -->
<!--   d3heatmap(pve_values, colors="Reds", Colv=FALSE,  yaxis_font_size = "9pt", Rowv=dendrogram()) -->
<!-- }) -->
<!-- ``` -->



<!-- TF exploration -->
<!-- ===================================== -->

<!-- Inputs {.sidebar} -->
<!-- ------------------------------------- -->
<!-- ```{r} -->
<!-- selectInput("gene2", label = h4("Selected gene"), -->
<!--     choices = gene_names, -->
<!--     selected = "PTPN14") -->
<!-- selectInput("cancer2", label = h4("Selected cancer"), -->
<!--     choices = names(TF_values), -->
<!--     selected = "BLCA") -->
<!-- selectInput("genelist2", label = h4("Selected list of genes"),  -->
<!--             choices = unique(unlist(lapply(TF_values, rownames))), -->
<!--                     multiple=TRUE, selected=unique(unlist(lapply(TF_values, rownames)))[1:5]) -->
<!-- ``` -->

<!-- Column {.tabset .tabset-fade} -->
<!-- ----------------------------------------------------------------------- -->
<!-- ### Gene symbol {data-padding=50} -->
<!-- ```{r} -->
<!-- gene2 <- reactive(as.character(input$gene2)) -->
<!-- renderD3heatmap({ -->
<!--   ## Read in data -->
<!--    #  sel <- lapply(TF_values, function(x) x[which(x$gene == "PTPN14"),])  -->

<!--   sel <- lapply(TF_values, function(x) x[which(x$gene == gene2()),])  -->
<!--   sel <- rbindlist(sel, fill=TRUE)[, cancer := rep(names(sel), vapply(sel, nrow, 0L))][] -->
<!--   sel <- as.data.frame(sel) -->
<!--   rownames(sel) <- sel$cancer -->
<!--   tf_all <- subset(sel, select=-c(gene, cancer)) -->
<!--   tf_all[is.na(tf_all)] <- 0 -->
<!--   if(length(which(colSums(tf_all) == 0))) tf_all <- tf_all[,-which(colSums(tf_all) == 0)] -->

<!--  pal <- colorRampPalette(c("#045A8D", "white", "red"))(9) -->
<!--  mi <- min(tf_all, na.rm = TRUE) -->
<!--  ma <- max(tf_all, na.rm = TRUE) -->
<!--  breaks <- c(seq(mi, -0.01, length=5), seq(0.01, ma, length=5)) -->
<!--  colorFunc <- col_bin(pal, bins = rescale(breaks)) -->
<!--  if(sum(rowSums(tf_all)==0))   tf_all <- tf_all[-which(rowSums(abs(sign(tf_all)))==0),] -->
<!--  if(sum(colSums(tf_all)==0)) tf_all <- tf_all[,-which(colSums(abs(sign(tf_all)))==0)] -->

<!--  d3heatmap(t(tf_all), Colv=TRUE,  yaxis_font_size = "9pt", xaxis_font_size = "0pt", -->
<!--            colors=colorFunc) -->
<!--  }) -->
<!-- ``` -->


<!-- ### Cancer {data-padding=50} -->


<!-- ```{r} -->
<!-- cancer2 <- reactive(as.character(input$cancer2)) -->
<!-- genelist2 <- reactive(input$genelist2) -->

<!-- renderD3heatmap({ -->
<!--  ## Read in data -->
<!--   # sel <- TF_values[["BRCA"]] -->
<!--   # sel <- sel[which(sel$gene %in% c("A2ML1", "AADAC", "AAADAT", "AASS", "A2M")),]  -->

<!--   sel <- TF_values[[cancer2()]] -->
<!--   sel <- sel[which(sel$gene %in% genelist2()),] -->

<!--   rownames(sel) <- sel$gene -->
<!--   tf_all <- subset(sel, select=-c(gene)) -->
<!--   tf_all[is.na(tf_all)] <- 0 -->
<!--   if(length(which(colSums(tf_all) == 0))) tf_all <- tf_all[,-which(colSums(tf_all) == 0)] -->

<!--  pal <- colorRampPalette(c("#045A8D", "white", "red"))(9) -->
<!--  mi <- min(tf_all, na.rm = TRUE) -->
<!--  ma <- max(tf_all, na.rm = TRUE) -->
<!--  breaks <- c(seq(mi, -0.01, length=5), seq(0.01, ma, length=5)) -->
<!--  colorFunc <- col_bin(pal, bins = rescale(breaks)) -->
<!--  d3heatmap(tf_all, Colv=TRUE,  yaxis_font_size = "9pt", xaxis_font_size = "0pt", -->
<!--            colors=colorFunc) -->
<!-- }) -->
<!-- ``` -->





<!-- ### Pan-cancer (2) {data-padding=50} -->
<!-- ```{r} -->

<!-- renderD3heatmap({ -->
<!--   ## Read in data -->
<!--   # pve <- lapply(cancerres, function(x) x[which(rownames(x) == "A1BG"),]) %>% -->
<!--   #   do.call("rbind", .) -->
<!--   # pve2 <- lapply(pve_values2, function(x) x[which(rownames(x) == "A2BP1"),]) %>% -->
<!--   #   do.call("rbind", .) -->
<!--   pve2 <- lapply(pve_values2, function(x) x[which(rownames(x) == gene()),]) %>% -->
<!--     do.call("rbind", .) -->
<!-- #  pve2 <- lapply(pve_values2, function(x) x[which(rownames(x) == "BRCA2"),]) %>% -->
<!-- #    do.call("rbind", .) -->
<!--  d3heatmap(pve2, colors="Reds", Colv=FALSE,  yaxis_font_size = "9pt") -->

<!-- }) -->
<!-- ``` -->




<!-- Wish list -->
<!-- ===================================== -->
<!-- - Couple heatmap with boxplots: https://plot.ly/r/shinyapp-linked-click/ -->
<!-- - Link to TCGA data, Ensembl info (https://plot.ly/r/shiny-coupled-hover-events/) -->


<!-- Test -->
<!-- =====================================  -->
<!-- Inputs {.sidebar} -->
<!-- ----------------------------------------------------------------------- -->

<!-- Identifying the genetic, epigenetic, and clinical drivers in gene expression from -->
<!-- the TCGA data. -->

<!-- ```{r} -->
<!-- selectInput("cancer2", label = h3("Cancer"),  -->
<!--     choices = cancer_choice,  -->
<!--     selected = "BRCA") -->
<!-- ``` -->

<!-- Row {data-height=650} -->
<!-- ----------------------------------------------------------------------- -->
<!-- ### Boxplots of percent variance explained (PVE), per source -->
<!-- ```{r} -->
<!-- cancer2 <- reactive(as.character(input$cancer2)) -->


<!-- output$boxplot <- renderPlotly({ -->
<!--   res <- cancerres[[cancer2()]] -->
<!--   res <- dplyr::select(res, starts_with("mirna"), starts_with("methyl"), mut, cna, G, resid) -->
<!--   res <- res / rowSums(res, na.rm=TRUE) -->
<!--   pve <- mutate(res, mirna = mirna_PC1 + mirna_PC2 + mirna_PC3 + mirna_PC4 + mirna_PC5) %>% -->
<!--     mutate(methyl = methyl_logit) %>% mutate(genetic = G) %>%  -->
<!--     mutate(residual = resid) %>% -->
<!--     dplyr::select(mut, cna, methyl, mirna, genetic, residual) -->
<!--   rownames(pve) <- rownames(res) -->
<!--   ## Remove pve that are 0 everywhere (correspond to RNA-seq with only 0 values) -->
<!--   pve <- pve[which(rowSums(pve) > 0),] -->
<!--   ## Replace NA's with 0 -->
<!--   pve[is.na(pve)] <- 0 -->
<!--   ## Format data for ggplot2 -->
<!--   pve_df <- data.frame(ID = rownames(pve), pve) -->
<!--   pve_long <- gather(pve_df, source, pve, -ID) -->
<!--   pve_long$source <- factor(pve_long$source, levels=colnames(pve)) -->
<!--   quants <- do.call("rbind", lapply(pve, summary)) -->
<!--   IQR <- (quants[,5] - quants[,2]) * 1.5 -->
<!--   lb <- quants[,2]- IQR -->
<!--   ub <- quants[,5]+IQR -->
<!--   pve_outlier <- rbind(pve_long %>% filter(source == "mut" & pve > ub["mut"]), -->
<!--                        pve_long %>% filter(source == "cna" & pve > ub["cna"]), -->
<!--                        pve_long %>% filter(source == "methyl" & pve > ub["methyl"]), -->
<!--                        pve_long %>% filter(source == "mirna" & pve > ub["mirna"]), -->
<!--                        pve_long %>% filter(source == "genetic" & pve > ub["genetic"]), -->
<!--                        pve_long %>% filter(source == "residual" & pve > ub["residual"])) -->
<!--     g2 <- ggplot() +  -->
<!--         geom_boxplot(data=pve_long, aes(x=source, y=pve, fill=source), outlier.color="white") +              geom_point(data=pve_outlier, aes(label=ID, x=source, y=pve)) + -->
<!--         theme_bw() +  -->
<!--     theme(legend.position="none") -->
<!--    ggplotly(g2, tooltip=c("ID", "source", "pve"), layerData=2) -->
<!-- }) -->
<!-- plotlyOutput("boxplot") -->
<!-- ``` -->

<!-- ### Another one -->
<!-- ```{r} -->
<!-- click_marker <- eventReactive(input$plotly_click, { -->
<!-- #  s <- input$plotly_click -->
<!--      s <- event_data("plotly_click") -->
<!--   return(s) -->
<!-- }) -->

<!-- data_for_chart <- reactive({ -->
<!--     res <- lapply(cancerres, function(x) x[which(rownames(x) == click_marker()),1:10]) %>% -->
<!--         do.call("rbind", .) -->
<!--     res <- res / rowSums(res, na.rm=TRUE) -->
<!--     pve <- mutate(res, mirna = mirna_PC1 + mirna_PC2 + mirna_PC3 + mirna_PC4 + mirna_PC5) %>% -->
<!--         mutate(methyl = methyl_logit) %>% mutate(genetic = G) %>% -->
<!--         mutate(residual = resid) %>% -->
<!--         dplyr::select(mut, cna, methyl, mirna, genetic, residual) -->
<!--     rownames(pve) <- rownames(res) -->
<!--     pve[is.na(pve)] <- 0 -->
<!--     return(pve) -->
<!-- }) -->

<!--  output$selection <- renderPrint({ -->
<!--    s <- click_marker() -->
<!--    if(length(s) == 0) { -->
<!--      "Click on an outlier in the boxpot to display a scatterplot" -->
<!--    } else { -->
<!--      cat("You selected: \n\n") -->
<!--      as.list(s) -->
<!--    } -->
<!--  })  -->
<!-- textOutput("selection") -->
<!-- plotlyOutput("heat") -->
<!-- ``` -->



<!-- ### PVE boxplots {data-padding=20} -->
<!-- ```{r} -->

<!-- renderPlot({ -->

<!-- ## Read in data -->
<!-- pve <- pve_values3[[cancer()]] -->
<!-- pve_df <- data.frame(ID = rownames(pve), pve) -->
<!-- pve_long <- gather(pve_df, source, pve, -ID) %>% ungroup() -->
<!-- pve_long$source <- factor(pve_long$source, levels=colnames(pve)) -->
<!-- quants <- do.call("rbind", lapply(pve, function(x) summary(na.omit(x)))) -->
<!-- IQR <- (quants[,5] - quants[,2]) * 1.5 -->
<!-- lb <- quants[,2]- IQR -->
<!-- ub <- quants[,5]+IQR -->
<!-- pve_outlier <- rbind(pve_long %>% filter(source == "mut" & pve > ub["mut"]), -->
<!--                      pve_long %>% filter(source == "cna" & pve > ub["cna"]), -->
<!--                      pve_long %>% filter(source == "methyl" & pve > ub["methyl"]), -->
<!--                      pve_long %>% filter(source == "mirna" & pve > ub["mirna"])) -->
<!-- #                     pve_long %>% filter(source == "genetic" & pve > ub["genetic"]), -->
<!-- #                     pve_long %>% filter(source == "residual" & pve > ub["residual"])) -->

<!--   g2 <- ggplot() +  -->
<!--     geom_boxplot(data=pve_long, aes(x=source, y=pve, fill=source), outlier.color="black") +               -->
<!--     geom_point(data=pve_outlier, aes(label=ID, x=source, y=pve)) + -->
<!--     scale_fill_brewer(palette = "Accent", direction=-1) + -->
<!--     theme_bw() +  -->
<!--     theme(legend.position="none") -->
<!-- #  ggplotly(g2, tooltip=c("ID", "source", "pve"), layerData=2, originalData=FALSE) -->
<!--   g2 -->
<!-- }, res=110) -->
<!-- ``` -->





<!-- ```{r} -->
<!-- plotlyOutput("heat") -->
<!-- output$heat <- renderPlotly({ -->
<!--   ## Read in data -->
<!--   pve0 <- pve_values3[[cancer()]] -->
<!--   pve0 <- sapply(pve0, function(x) as.numeric(as.character(x))) -->
<!--   set.seed(12345) -->
<!--   pve0[which(is.na(pve0)==TRUE)] <- 0 -->
<!--   set.seed(12345) -->
<!--   km <- kmeans(pve0, 20)$cluster -->
<!--   ## Average profiles over K-means clusters -->
<!--   ## Format data for ggplot2 -->
<!--   pve_km <- rowsum(pve0, group=km) / as.vector(table(km)) -->
<!--   rownames(pve_km) <- paste("Cluster", 1:20) -->

<!--   pve_km <- t(pve_km) -->
<!--   colnames(pve_km) <- factor(colnames(pve_km), levels=paste("Cluster", 1:20)) -->
<!--   plot_ly(x = colnames(pve_km), y = rownames(pve_km), z = pve_km,  -->
<!--             key = pve_km, type = "heatmap", colorscale="RdBu") %>% -->
<!--       layout(xaxis = list(title = ""),  -->
<!--              yaxis = list(title = "")) -->
<!-- }) -->
<!-- ``` -->



<!-- ```{r} -->
<!-- verbatimTextOutput("selection") -->
<!-- output$selection <- renderPrint({ -->
<!--   s <- event_data("plotly_click") -->
<!--   if (length(s) == 0) { -->
<!--       "Click on a cell in the heatmap to display a scatterplot" -->
<!--     } else { -->
<!--       cat("You selected: \n\n") -->
<!--       c(s$x, s$y, s$z, "\n") -->
<!--     } -->
<!-- }, width=120) -->
<!-- ``` -->

<!-- renderPlotly({ -->
<!--   ## Read in data -->
<!--   pve0 <- pve_values3[[cancer()]] -->
<!--   pve0 <- sapply(pve0, function(x) as.numeric(as.character(x))) -->
<!--   set.seed(12345) -->
<!--   pve0[which(is.na(pve0)==TRUE)] <- 0 -->
<!--   set.seed(12345) -->
<!--   km <- kmeans(pve0, 20)$cluster -->
<!--   ## Average profiles over K-means clusters -->
<!--   ## Format data for ggplot2 -->
<!--   pve_km <- rowsum(pve0, group=km) / as.vector(table(km)) -->
<!--   rownames(pve_km) <- paste("Cluster", 1:20) -->

<!--   plot_ly(x = colnames(pve_km), y = rownames(pve_km), z = pve_km,  -->
<!--             key = pve_km, type = "heatmap") %>% -->
<!--       layout(xaxis = list(title = ""),  -->
<!--              yaxis = list(title = "")) -->
<!--   # pve_df <- data.frame(ID = rownames(pve_km), pve_km) -->
<!--   # pve_long <- gather(pve_df, source, pve0, -ID) %>% ungroup() -->
<!--   # pve_long$source <- factor(pve_long$source, levels=colnames(pve_km)) -->
<!--   # pve_long$ID <- factor(pve_long$ID, levels=paste("Cluster", 1:20)) -->
<!--   # levels(pve_long$ID) <- paste0(levels(pve_long$ID), "\n(", as.vector(table(km)), " genes)") -->

<!--   g3 <- ggplot(pve_long, aes(x=ID, fill=source, y=pve0, order=source))  +  -->
<!--     geom_bar(stat="identity") +  -->
<!--     coord_flip() + -->
<!--     theme_classic() + scale_fill_brewer(palette = "Accent", direction=-1) +  -->
<!--     theme(axis.title.y=element_blank())  + xlab("Variance components") -->
<!--   g3 -->
<!-- #  ggplotly(g3, tooltip=c("ID", "source")) -->
<!-- }, res=110) -->
<!-- ``` -->



<!-- ### K-means clusters {data-padding=20} -->
<!-- ```{r} -->
<!-- renderPlot({ -->
<!--   ## Read in data -->
<!--   #  sel <- all_values[cancer == "BRCA", .(gene, mut, cna, mirna, TF, methyl, genetic, residual)] -->
<!--   sel <- all_values[cancer == cancer(), .(gene, mut, cna, mirna, TF, methyl, genetic, residual)] -->

<!--   sel <- as.data.frame(sel, stringsAsFActors=FALSE) -->
<!--   rownames(sel) <- sel$gene -->
<!--   pve_values <- sel[,-1] -->
<!--   pve_values[is.na(pve_values)] <- 0 -->
<!--   pve_values <- na.omit(pve_values) -->

<!--   pve_values <- pve_values / rowSums(pve_values) -->

<!--   tmp <- clr(pve_values) -->
<!--   attributes(tmp) <- NULL -->
<!--   tmp <- matrix(tmp, nrow = nrow(pve_values), ncol = ncol(pve_values)) -->
<!--   rownames(tmp) <- rownames(pve_values) -->
<!--   colnames(tmp) <- colnames(pve_values) -->

<!--   set.seed(12345) -->
<!--   K <- 20 -->
<!--   km <- kmeans(tmp, K)$cluster -->

<!--   ## Average profiles over K-means clusters, format data -->
<!--   pve_km <- rowsum(tmp, group=km) / as.vector(table(km)) -->
<!--   rownames(pve_km) <- paste0("Cluster ", 1:K, " (", table(km), " genes)") -->

<!--   pve_km <- t(pve_km) -->
<!--   pheatmap(pve_km[-7,], color = colorRampPalette(c("white", "yellow", "red"), bias=2)(100), -->
<!--            cluster_cols=TRUE) -->
<!-- }, res=105) -->
<!-- ``` -->



<!-- ### K-means cluster lists {data-padding=20} -->

<!-- ```{r} -->
<!-- selectInput("kmc", label = h3("K-means cluster"), -->
<!--     choices = 1:20, -->
<!--     selected = 1) -->

<!-- kmc <- reactive({as.numeric(input$kmc)}) -->

<!-- renderPrint({ -->
<!--   ## Read in data -->
<!--   #  sel <- all_values[cancer == "BRCA", .(gene, mut, cna, mirna, TF, methyl, genetic)] -->
<!--   sel <- all_values[cancer == cancer(), .(gene, mut, cna, mirna, TF, methyl, genetic)] -->

<!--   sel <- as.data.frame(sel, stringsAsFActors=FALSE) -->
<!--   rownames(sel) <- sel$gene -->
<!--   pve_values <- sel[,-1] -->
<!--   pve_values[is.na(pve_values)] <- 0 -->
<!--   set.seed(12345) -->
<!--   K <- 20 -->
<!--   km <- kmeans(pve_values, K)$cluster -->

<!--   ind <- rownames(pve_values)[which(km == kmc())] -->
<!--   data.frame(GeneSymbol = ind) -->
<!-- }) -->
<!-- ``` -->



<!-- ### TF overview  {data-padding=70} -->
<!-- Hover over the heatmap to see sparse principal component loading for each TF, weighted by its linear model coefficient. -->

<!-- ```{r} -->
<!-- renderD3heatmap({ -->
<!--   # sel <- TF_values[gene == "PTPN14"] -->
<!--   sel <- TF_values[gene == gene()] -->

<!--   sel_list <- vector("list", nrow(sel)) -->
<!--   names(sel_list) <- sel$cancer -->
<!--   for(i in seq_len(nrow(sel))) { -->
<!--     sel_list[[i]] <- data.frame(sel[i,-c(1:3)], check.names=FALSE) -->
<!--     colnames(sel_list[[i]]) <- strsplit(TF_key[as.numeric(sel[i,3]),2], ";")$tfs -->
<!--   } -->
<!--   sel_list <- rbindlist(sel_list, idcol="cancer", fill=TRUE) -->
<!--   sel_list <- data.frame(sel_list, stringsAsFactors=FALSE, check.names=FALSE) -->
<!--   rownames(sel_list) <- sel_list$cancer -->
<!--   sel_list <- sel_list[,-1] -->
<!--   sel_list[is.na(sel_list)] <- 0 -->

<!--   sel_list <- sel_list[,which(colSums(abs(sign(sel_list))) > 0)] -->

<!--   # sel <- lapply(TF_values, function(x) x[which(x$gene == gene()),]) -->
<!--   # sel <- rbindlist(sel, fill=TRUE)[, cancer := rep(names(sel), vapply(sel, nrow, 0L))][] -->
<!--   # sel <- as.data.frame(sel) -->
<!--   # rownames(sel) <- sel$cancer -->
<!--   # tf_all <- subset(sel, select=-c(gene, cancer)) -->
<!--   # tf_all[is.na(tf_all)] <- 0 -->
<!--   # if(length(which(colSums(tf_all) == 0))) tf_all <- tf_all[,-which(colSums(tf_all) == 0)] -->

<!--  pal <- colorRampPalette(c("#045A8D", "white", "red"))(9) -->
<!--  mi <- min(sel_list, na.rm = TRUE) -->
<!--  ma <- max(sel_list, na.rm = TRUE) -->
<!--   breaks <- c(seq(mi, -0.01, length=10), seq(0.01, ma, length=10)) -->
<!--  colorFunc <- col_bin(pal, bins = rescale(breaks)) -->
<!--  d3heatmap(as.matrix(sel_list), Colv=TRUE,  yaxis_font_size = "9pt", xaxis_font_size = "0pt", -->
<!--            colors=colorFunc) -->
<!-- }) -->
<!-- ``` -->


<!-- ### miRNA overview {data-padding=70} -->
<!-- Hover over the heatmap to see sparse principal component loading for each miRNA, weighted by its linear model coefficient. -->

<!-- ```{r} -->
<!-- renderD3heatmap({ -->
<!--    # sel <- mir_values[gene == "PTPN14"] -->
<!--   sel <- mir_values[gene == gene()] -->

<!--   sel_list <- vector("list", nrow(sel)) -->
<!--   names(sel_list) <- sel$cancer -->
<!--   for(i in seq_len(nrow(sel))) { -->
<!--     sel_list[[i]] <- data.frame(sel[i,-c(1:3)], check.names=FALSE) -->
<!--     colnames(sel_list[[i]]) <- strsplit(mir_key[as.numeric(sel[i,2]),2], ";")$mirs -->
<!--   } -->
<!--   sel_list <- rbindlist(sel_list, idcol="cancer", fill=TRUE) -->
<!--   sel_list <- data.frame(sel_list, stringsAsFactors=FALSE, check.names=FALSE) -->
<!--   rownames(sel_list) <- sel_list$cancer -->
<!--   sel_list <- sel_list[,-1] -->
<!--   sel_list[is.na(sel_list)] <- 0 -->

<!--   sel_list <- sel_list[,which(colSums(abs(sign(sel_list))) > 0)] -->

<!--  pal <- colorRampPalette(c("#045A8D", "white", "red"))(9) -->
<!--  mi <- min(sel_list, na.rm = TRUE) -->
<!--  ma <- max(sel_list, na.rm = TRUE) -->
<!--   breaks <- c(seq(mi, -0.01, length=10), seq(0.01, ma, length=10)) -->
<!--  colorFunc <- col_bin(pal, bins = rescale(breaks)) -->
<!--  d3heatmap(as.matrix(sel_list), Colv=TRUE,  yaxis_font_size = "9pt", xaxis_font_size = "0pt", -->
<!--            colors=colorFunc) -->
<!-- }) -->
<!-- ``` -->




<!-- Genome-wide overview -->
<!-- =====================================  -->

<!-- Inputs {.sidebar} -->
<!-- ----------------------------------------------------------------------- -->
<!-- ```{r} -->
<!-- selectInput("cancersub2", label = h5("Cancer subset"), -->
<!--     choices = as.character(cancer_choice), -->
<!--     selected = cancer_choice, multiple=TRUE) -->

<!-- cancersubset2 <- reactive({as.character(input$cancersub2)}) -->
<!-- ``` -->

<!-- Column {.tabset} -->
<!-- ----------------------------------------------------------------------- -->

<!-- <!-- ### Boxplots {data-padding=50} --> 

<!-- <!-- ```{r} --> 
<!-- <!-- renderPlot({ --> 
<!-- <!--   # sel <- all_values[cancer %in% c("BRCA", "BLCA")] --> 
<!-- <!--   sel <- all_values[cancer %in% cancersubset2()] --> 
<!-- <!--   varcomp <- sel %>% select(-starts_with("mean"), -heritability) --> 
<!-- <!--   varcomp_long <- gather(varcomp, source, varcomp, -cancer, -gene) --> 
<!-- <!--   varcomp_long[is.na(varcomp_long)] <- 0 --> 
<!-- <!--   varcomp_long$source <- factor(varcomp_long$source, levels=colnames(varcomp)[-c(1:2)]) --> 

<!-- <!--   g1 <- ggplot(varcomp_long, aes(x=source, y=varcomp)) + --> 
<!-- <!--     geom_boxplot(aes(fill=source)) + --> 
<!-- <!--     theme_bw() + --> 
<!-- <!--     scale_fill_brewer(palette = "Accent", direction=-1) + --> 
<!-- <!--     theme(legend.position="none") + --> 
<!-- <!--     ylab("Variance component") + xlab("Molecular source") + --> 
<!-- <!--     ylim(0,1.25) + --> 
<!-- <!--     facet_wrap(~cancer) + --> 
<!-- <!--     theme(axis.text.x  = element_text(angle=45, vjust=0.5)) --> 
<!-- <!--   print(g1) --> 
<!-- <!-- }) --> 
<!-- <!-- ``` --> 


<!-- ### Violin plots {data-padding=50} -->

<!-- ```{r} -->
<!-- renderPlot({ -->
<!--   # sel <- all_values[cancer %in% c("BRCA", "BLCA")] -->
<!--   sel <- all_values[cancer %in% cancersubset2()] -->
<!--   varcomp <- sel %>% dplyr::select(-starts_with("mean"), -heritability) -->
<!--   varcomp[is.na(varcomp)] <- 0 -->

<!--  #  par(mar=c(3,4,3,2)) -->
<!--  #  auto.layout(length(cancersubset2())) -->
<!--  #  colvec <- brewer.pal(7, "Accent") -->
<!--  #  for(can in cancersubset2()) { -->
<!--  #    tmp <- as.data.frame(varcomp[cancer==can, .(mut, cna, methyl, mirna, TF, genetic, residual)]) -->
<!--  #    plot(0,0,type="n",xlim=c(0.5,7.5), ylim=c(0,1.25),xaxt = 'n', xlab ="", -->
<!--  #         ylab = "Variance component",  main =can) -->
<!--  #    for (i in 1:7) { -->
<!--  #      if(sum(tmp[,i]) == 0) next; -->
<!--  # #     vioplot(tmp[,i], at = i, add = TRUE, col = colvec[i]) -->
<!--  #      beanplot(tmp[,i], at = i, add = TRUE, col = colvec[i]) -->
<!--  #    } -->
<!--  #    axis(side=1, at=1:7,labels=c("mutation", "CNA", "methylation", "miRNA", "TF", "genetic", -->
<!--  #                                 "residual")) -->
<!--  #    # vioplot(tmp$mut, tmp$cna, tmp$methyl, tmp$mirna, tmp$TF, tmp$genetic, tmp$residual, -->
<!--  #    #         names=c("mutation", "CNA", "methylation", "miRNA", "TF", "genetic", "residual"), -->
<!--  #    #         col=brewer.pal(7, "Accent"), rectCol="gray") -->
<!--  #    # mtext(side=3, can, 1) -->
<!--  #  } -->

<!--   varcomp_long <- gather(varcomp, source, varcomp, -cancer, -gene, -chr, -strand, -tss) -->
<!--   varcomp_long[is.na(varcomp_long)] <- 0 -->
<!--   varcomp_long$source <- factor(varcomp_long$source, levels=colnames(varcomp)[-c(1:2)]) -->
<!-- #  varcomp_long$varcomp <- as.numeric(varcomp_long$varcomp) -->


<!--    g2 <- ggplot(varcomp_long, aes(x=source, y=varcomp)) + -->
<!--      geom_violin(aes(fill=source),scale="width") + -->
<!--      theme_bw() + -->
<!--      theme(legend.position="none") + -->
<!--      ylab("Variance component") + xlab("Molecular source") + -->
<!--      ylim(0,1.25) + -->
<!--      facet_wrap(~cancer) + -->
<!--      theme(axis.text.x  = element_text(angle=45, vjust=0.5)) -->
<!--    print(g2) -->
<!-- }) -->

<!-- ``` -->


<!-- ### Correlation plots  {data-padding=20, data-width=600} -->

<!-- ```{r} -->
<!-- selectInput("sourcee", label = h5("Molecular source"), -->
<!--     choices = c("mutation", "CNA", "methylation", "miRNA", "TF", "genetic"), -->
<!--     selected = "TF") -->

<!-- molsource <- reactive({as.character(input$sourcee)}) -->
<!-- #  -->
<!-- # downloadHandler(filename = function() {paste('correlation-', Sys.Date(), '.pdf', sep='')}, -->
<!-- #                 content = function(file) { -->
<!-- #                   pdf(file, width=5.5, height=6.5, onefile=FALSE) -->
<!-- #                   corrcalc <- cor_list[[molsource()]] -->
<!-- #                 index <- which(rownames(corrcalc) %in% cancersubset2()) -->
<!-- #   corrcalc <- corrcalc[index,index] -->
<!-- #   paletteLength <- 50 -->
<!-- #   myColor <- colorRampPalette(c("#045A8D", "white", "red"))(paletteLength) -->
<!-- #   myBreaks <- c(seq(-1, 0, length.out=ceiling(paletteLength/2) + 1), -->
<!-- #                 seq(1/paletteLength, 1, length.out=floor(paletteLength/2))) -->
<!-- #   pheatmap(corrcalc, color=myColor, breaks=myBreaks, main = "") -->
<!-- #                   dev.off() -->
<!-- #                   },  -->
<!-- #                 contentType="application/pdf") -->

<!-- renderPlot({ -->
<!--   corrcalc <- cor_list[[molsource()]] -->
<!--   index <- which(rownames(corrcalc) %in% cancersubset2()) -->
<!--   corrcalc <- corrcalc[index,index] -->
<!--   paletteLength <- 50 -->
<!--   myColor <- colorRampPalette(c("#045A8D", "white", "red"))(paletteLength) -->
<!--   myBreaks <- c(seq(-1, 0, length.out=ceiling(paletteLength/2) + 1), -->
<!--                 seq(1/paletteLength, 1, length.out=floor(paletteLength/2))) -->
<!--   pheatmap(corrcalc, color=myColor, breaks=myBreaks, main = "") -->
<!-- }, height=550, width=650) -->
<!-- ``` -->




<!-- Genome-wide pan-cancer -->
<!-- =====================================  -->

<!-- Inputs {.sidebar} -->
<!-- ----------------------------------------------------------------------- -->
<!-- ```{r} -->
<!-- selectInput("molsource", label = h6("Select a source of molecular variation to visualize:"), -->
<!--             choices = c("mut", "cna", "methyl", "mirna", "TF", "genetic"), -->
<!--             selected = "TF", multiple=FALSE) -->

<!-- #actionButton("go", "Make pan-cancer plots!") -->
<!-- ``` -->

<!-- <br> -->
<!-- <br> -->
<!-- <br> -->


<!-- This plot represents the first two principal components for the genomewide variance components from a user-selected source of molecular variation for each of the 17 cancers included in this study. This allows a visualization of how cancers cluster with one another according to global patterns of how gene expression variation is decomposed into different molecular components. -->


<!-- Column  -->
<!-- ----------------------------------------------------------------------- -->
<!-- ### Pan-cancer PCA {data-padding=25} -->

<!-- ```{r} -->

<!-- molsource <- reactive(as.character(input$molsource)) -->

<!-- renderPlot({ -->
<!--   tmp <- all_values[,c("gene", "cancer", molsource())] -->
<!--   tmp2 <- spread(tmp, key="cancer", value=molsource()) %>% dplyr::select(-gene) -->
<!--   tmp2 <- tmp2[which(rowSums(is.na(tmp2)) == 0),] -->
<!--   pca_analysis <- pca(t(tmp2)) -->
<!--   pp <- plotIndiv(pca_analysis, title="", cex=5)$graph  -->
<!--   pp + theme(plot.margin=margin(t=0.5,r=0.5,l=0.5,b=1.5, unit="in")) -->
<!-- }) -->
<!-- ``` -->